### PR TITLE
Restructure key4hep-bundle-package

### PR DESCRIPTION
**Pull request by vvolkl**
* general update for consistency and formatting
* ~add `rolling` version in addition to "nightly" and numbered versions~
  - use dates as versions for the stack
* add some general purpose generators
* ~add the concretizer workarounds from packages.yaml~

todos:
* ~add pinned version numbers for 0.1 for ilcsoft packages~
* ~think about how to proceed with `packages.yaml` vs `+concretizer hints`~
     - unfortunately, doing this in the recipe only works for versions, not for variants

EDIT: ~I propose to postpone finding a good way to fix the version numbers to a future PR.~

### CODE REVIEW

File: packages/key4hep-stack/package.py [hunk 1]
```diff
@@ -4,84 +4,238 @@
...
+    ##################### concretizer hints ###############
+    #######################################################
+    depends_on('boost +python', when="+concretizer-hints")
+    # avoid building doxygen
+    depends_on('range-v3 ~doc', when="+concretizer-hints")
+    # avoid building mariadb
+    depends_on('marlin ~lccd', when="+concretizer-hints")
+    # avoid building llvm
+    depends_on('mesa ~llvm+glx+opengl', when="+concretizer-hints")
+    depends_on('dd4hep +geant4', when="+concretizer-hints")
```
Comment (andresailer) [inline packages/key4hep-stack/package.py:226]:
```suggestion
    depends_on('dd4hep +geant4 +lcio', when="+concretizer-hints")
```
Or how is the lcio variant injected at the moment?

Comment (vvolkl) [reply packages/key4hep-stack/package.py:226]: good catch!

**Comment by vvolkl** [comment]
I think this is now a reasonably good way of versioning the stack:

The bundle package declares only general dependencies without versions. One exception: `key4hep-stack@master` will try to build  the master branch of almost all direct dependencies. This can be used for nighttly builds. Note that spack does not resolve branches, so nightly builds will have to remove the installs  or install to a different directory (improving this is WIP in upstream spack).

*Any other version than `master`* will build the latest stable version defined in the recipe for each package. There is no need to add those version to the recipe, and I propose to use dates, similar to ilcsoft.
Example:

`spack install key4hep-stack@2020-10-06`

The versions used in this install can be extracted by adding this install to an environment and generating a lockfile: https://spack-tutorial.readthedocs.io/en/latest/tutorial_environments.html

More specialised installs where versions are picked manually can be created by writing a `package.yaml` like

```
key4hep-stack:
    version: [v1.0]
  fccsw:
    version: [v0.12]
  guinea-pig:
    version: [1.2.2rc]
...

```
and ideally adding the file to this repo.
The versions for these can be numbers or a descriptive keyword, again, no need to add them to the recipe.

File: packages/key4hep-stack/package.py [hunk 2]
```diff
@@ -1,90 +1,225 @@
...
+    variant('devtools', default=True,
+            description="add tools necessary for software masterment to the stack")
+    variant('generators', default=True,
+            description="add some standalone generators to the stack")
+    variant('fccsw', default=True,
+            description="build fcc packages")
+    variant('ilcsoft', default=True,
+            description="build ilcsoft packages")
+    variant('bootstrap', default=True,
+            description="install some spack setup tools")
```
Comment (tmadlener) [inline packages/key4hep-stack/package.py:32]: I really like this idea of having different variants in general. However, I am wondering if it scales in the end, especially in the cases, where the different variants start to overlap. As long as there are no overlaps between the variants we can simply do

```python
depends_on('fcc_package', when='+fccsw')
depends_on('ilc_package', when='+ilcsoft')
```

but as far as I know there is no easy way to say that a given package should be built when either of the variants is active. It seems to be possible to do something like this:
```python
depends_on('package', when='+fccsw')
depends_on('package', when='+ilcsoft')
```

This gets even more verbose if we want to tie in the `@master` "variant", as we would have to define all the versions a second time with `@master`.

Or do we simply consider packages that would be shared by more than one variant to be always built?

File: packages/key4hep-stack/package.py [hunk 3]
```diff
@@ -1,90 +1,225 @@
...
+    #######################################################
+    depends_on("madgraph5amc", when="+generators")
+    depends_on("herwigpp", when="+generators")
+    # todo: investigate build failure with newer versions
+    depends_on("lhapdf@6.2.3", when="+generators")
+
+
+
+    ############################### ilcsoft ###############
+    #######################################################
```
Comment (tmadlener) [inline packages/key4hep-stack/package.py:64]: These should all get the `+ilcsoft` variant, right?

File: packages/key4hep-stack/package.py [hunk 4]
```diff
@@ -1,90 +1,225 @@
...
+    depends_on("raida")
+    depends_on('raida@master', when='@master')
+
+    depends_on('sio')
+    depends_on('sio@master', when='@master')
+
+
+    ############################### fccsw #################
+    #######################################################
+    depends_on("fccsw")
```
Comment (tmadlener) [inline packages/key4hep-stack/package.py:195]: And these should all get the `+fccsw` variant, I think.

File: packages/key4hep-stack/package.py (hunk 2 above)
Comment (vvolkl) [reply packages/key4hep-stack/package.py:30]: Yes, I'm actually not sure anymore these variants are needed at all -- I propose to move this issue to a new PR.

> Or do we simply consider packages that would be shared by more than one variant to be always built?

Yes, in that case I would have just dropped the `when`.

**Comment by vvolkl** [comment]
Now allows to  add a "dynamic" version `develop-<commithash>` for the latest commit in the repository (looked up on the fly). This finally should allow making efficient nightly builds.

**Comment by vvolkl** [comment]
Seems to work in my test. I'll merge this to setup an automated nightly build, but feel free to comment to request changes in further PRs.

**Pull request merged**